crowbar_framework/app/models/service_object.rb

```diff
@@ -1540,6 +1524,79 @@ def wait_for_chef_daemons(node_list)
 
   private
 
+  THREAD_POOL_SIZE = 20
+
+  def release_chef_locks(locks)
+    return if locks.empty?
+
+    queue = Queue.new
+    locks.each { |l| queue.push l }
+
+    workers = (1...THREAD_POOL_SIZE).map do
```

Comment:
Gents, my ruby knowledge is also not the best, I guess the statement:

```ruby
      loop do
        begin
          node = queue.pop(true)
        rescue ThreadError
          puts "ERROR"
          break
        end
```
Does the trick, so each thread tries to pop, and if the queue is empty, it raises an error. 
that's how the queue is emptied @aplanas Having said that it does not communicate the intent too well, and it's quite obscure. See this "magic":

```ruby
2.1.2 :001 > q = Queue.new
 => #<Thread::Queue:0x00000001194730>
2.1.2 :002 > q.push 1
 => #<Thread::Queue:0x00000001194730>
2.1.2 :003 > q.pop
 => 1
2.1.2 :004 > q.pop
fatal: No live threads left. Deadlock?
        from (irb):4:in `pop'
        from (irb):4
        from /home/matelakat/.rvm/rubies/ruby-2.1.2/bin/irb:11:in `<main>'
```

Now the interesting part comes when we try to do another pop after that:
```ruby
2.1.2 :005 > q.pop
```
It blocks the calling thread (as it would be expected)

I might not be well educated enough to find an explanation for this behaviour.